### PR TITLE
Customization options

### PR DESCRIPTION
**Pull request by Bubobubobubobubo**
This PR is rather large and unfocused so I will keep it as a draft for now. Some help is necessary to review the code in order to tame the remaining ugly bits. I don't know much about React and my frontend skills are shallow. The PR is adding some much needed customization options to Flok:

- **Fonts:** JetBrains, Monaco, Courier, MonoCraft, BigBlue Terminal, JGS, Steps Mono, IBM Plex Mono, Inconsolata.
- **Themes:** Dracula, Monokai Dimmed, Gruvbox Dark, Ayu Dark, Tokyo Night, Nord.

It also refactors the keybinding based CodeMirror extension system that Flok was using. Now, options are accessed through the menu:

- **Vim Mode:** toggle / untoggle the Vim Editor keybindings
- **Wrap Lines:** wrap the line when it reaches the end of the screen
- **Line Numbers:** show or hide line numbers

**EDIT:** well, lots of typing errors to fix.

### CODE REVIEW

**Comment by munshkr** [comment]
> I fixed the typing errors in the code I've written. Did I break something in the build process?

Sorry I've been a bit busy with life and work.. I'll take a look this weekend and help if I can.

**Comment by tmhglnd** [comment]
Hey! I tried to check out this branch but I get a build error when running `npm build`:

```
 ✖  @flok-editor/session:build
       > @flok-editor/session@1.1.0 build
       > tsc
       
       ../../node_modules/y-websocket/dist/src/y-websocket.d.ts(116,28): error TS1479: The current file is a CommonJS module whose imports will produce 'require' calls; however, the referenced file is an ECMAScript module and cannot be imported with 'require'. Consider writing a dynamic 'import("lib0/observable")' call instead.
       ../../node_modules/y-websocket/dist/src/y-websocket.d.ts(117,20): error TS1479: The current file is a CommonJS module whose imports will produce 'require' calls; however, the referenced file is an ECMAScript module and cannot be imported with 'require'. Consider writing a dynamic 'import("yjs")' call instead.
       ../../node_modules/y-websocket/dist/src/y-websocket.d.ts(118,36): error TS1479: The current file is a CommonJS module whose imports will produce 'require' calls; however, the referenced file is an ECMAScript module and cannot be imported with 'require'. Consider writing a dynamic 'import("y-protocols/awareness")' call instead.
       ../../node_modules/y-websocket/dist/src/y-websocket.d.ts(119,27): error TS1479: The current file is a CommonJS module whose imports will produce 'require' calls; however, the referenced file is an ECMAScript module and cannot be imported with 'require'. Consider writing a dynamic 'import("lib0/encoding")' call instead.
       ../../node_modules/y-websocket/dist/src/y-websocket.d.ts(120,27): error TS1479: The current file is a CommonJS module whose imports will produce 'require' calls; however, the referenced file is an ECMAScript module and cannot be imported with 'require'. Consider writing a dynamic 'import("lib0/decoding")' call instead.
       lib/session.ts(257,55): error TS2339: Property 'closed' does not exist on type 'WebrtcProvider'.
       npm ERR! Lifecycle script `build` failed with error: 
       npm ERR! Error: command failed 
       npm ERR!   in workspace: @flok-editor/session@1.1.0 
       npm ERR!   at location: /Users/timohoogland/Downloads/flok-add-vim-mode/packages/session 
       npm notice 
       npm notice New minor version of npm available! 10.2.3 -> 10.8.0
       npm notice Changelog: <https://github.com/npm/cli/releases/tag/v10.8.0>
       npm notice Run `npm install -g npm@10.8.0` to update!
       npm notice 
       

 ——————————————————————————————————————————————————————————————————————————————————————————————————

 >  Lerna (powered by Nx)   Ran target build for 8 projects (6s)
 
    ✔    4/5 succeeded [0 read from cache]
 
    ✖    1/5 targets failed, including the following:
         - @flok-editor/session:build
```

I'm not so sure what to do to fix it.

**Comment by Bubobubobubobubo** [comment]
Yes I remember running a command that might have updated packages. The error has something to do with something unrelated to the changes: `Property 'closed' does not exist on type 'WebrtcProvider'`.

**Comment by munshkr** [comment]
@Bubobubobubobubo @tmhglnd yes, apparently a lot of packages were upgraded automatically at some point, including typescript, which causes that error. I reverted the changes on package-lock.json and re-installed the dependencies, and it's building again. Will look into that build problem later in another issue...

This looks great! I'm going to make a few more edits if you don't mind @Bubobubobubobubo

- [ ] Try to show Theme/Font options in a different way inside the Command dialog
- [ ] Show "Enable/Disable ..." in menu options based on its state
- [ ] Bug: Settings are not being stored on local storage anymore?
- [ ] Add a new option for enabling/disabling current line highlight (I prefer to keep it disabled by default, like we had before)
- [ ] Include old theme One Dark too

**Comment by Bubobubobubobubo** [comment]
Thank you for taking a look! Sorry about the auto-upgrade mishap. About customization, we could save a single object in localStorage to facilitate passing the configuration around in different components. I wasn't sure how localStorage was handled so I.. did nothing :smile:.

I can take a look and test if needed for the following commits :)

**Comment by tmhglnd** [comment]
Hey! I just briefly tried this. Some initial notes:

1. I don't see the names anymore? Maybe something broke there?
3. The theme's don't have a semi-transparent black background anymore. Making them very hard to read when Hydra visuals are being used.
2. The fontweight in the previous version was more bold/strong. I personally liked that for more visibility, dunnow if this should be an option or something we just put back in.
4. I like the Line-numbers and Word-wrapping features! In a previous commit I introduces `shift ctrl L` for enabling line-numbers ans `shift ctrl W` for word-wrap. Is this something we should put back in too?
5. I think word-wrap `on` should be the default maybe? At least I wouldn't mind.

**Comment by munshkr** [comment]
> Hey! I just briefly tried this. Very nice! 😎 Some initial notes:
> 
> 1. I don't see the names anymore? Maybe something broke there?
> 2. The theme's don't have a semi-transparent black background anymore. Making them very hard to read when Hydra visuals are being used.
> 3. The fontweight in the previous version was more bold/strong. I personally liked that for more visibility, dunnow if this should be an option or something we just put back in.
> 6. I also see this `..........` dotted line below the editor where i am currently typing in.

Most of these issues are related to the fact that I customized the One Dark theme a lot, so I guess we need to apply those on top of the selected theme.

> 4. I like the Line-numbers and Word-wrapping features! In a previous commit I proposed `shift-ctrl-L` for enabling line-numbers ans `shift-ctrl-W` for word-wrap. Is this something we should put back in too?

Agree, I think those are sensible shortcuts. I'll add a "Key Shortcuts" dialog soon, so this could be further customized.

> 5. I think word-wrap `on` should be the default maybe? At least I wouldn't mind.

Makes sense!

**Comment by tmhglnd** [comment]
I have a fix for **2**, removing all themes default background and replacing it by the semi-transparant black background that is not full editor but only where the text is.

I've also included some additional themes that I think ook nice.

I've also addressed **3**, making the default font-weight 500.

I have to up my github skills I think, because I'm not sure what now the best way is to get my commits to be in this fork. Should I do a pull request to this fork that then eventually will be merged with the main branch? (I guess so). In the meantime you can already check the things I'm doing here: https://github.com/tmhglnd/flok/tree/add-vim-mode

**Comment by Bubobubobubobubo** [comment]
You might be able to push to this branch by adding the repo it is based on as a remote.  Not a Git wizard myself 🥹

**Comment by tmhglnd** [comment]
> You might be able to push to this branch by adding the repo it is based on as a remote. Not a Git wizard myself 🥹

I would need permission for that, now I get this error: `remote: Permission to Bubobubobubobubo/flok.git denied to tmhglnd.` But I think I could make a pull request here and then you merge it with your fork. I'll look into it tomorrow!

**Comment by tmhglnd** [comment]
I made a pull request: https://github.com/Bubobubobubobubo/flok/pull/3 

Some extra things I updated:

- I normalized the font sizes in the @font-face with size-adjust: so there's less differences in font size when you change between them.

File: packages/web/src/components/editor.tsx [hunk 1]
```diff
@@ -30,50 +48,6 @@ const langExtensionsByLanguage: { [lang: string]: any } = {
...
-    color: "white",
-  },
-  "&.cm-focused": {
-    outline: "none",
-  },
-  ".cm-selectionBackground": {
-    backgroundColor: "rgba(255, 0, 255, 0.5) !important",
-    opacity: 0.5,
-  },
-  ".cm-ySelectionInfo": {
```
Comment (munshkr) [inline packages/web/src/components/editor.tsx:64]: @tmhglnd  I think this `.cm-ySelectionInfo` is related to the "username" cursor

File: packages/web/src/components/editor.tsx [hunk 2]
```diff
@@ -30,50 +48,6 @@ const langExtensionsByLanguage: { [lang: string]: any } = {
...
-    background: "transparent",
-    fontFamily: `Inconsolata`,
-    fontSize: "16px",
-    color: "white",
-    fontWeight: 600,
-  },
-  "& .cm-scroller": {
-    fontFamily: `Inconsolata`,
-    paddingLeft: "2px !important",
-    minHeight: "100vh",
```
Comment (munshkr) [inline packages/web/src/components/editor.tsx:44]: and I think this minHeight fixes the "......." outline that appears just below your cursor, it should expand the editor div height to full screen height.

File: packages/web/src/components/editor.tsx (hunk 1 above)
Comment (tmhglnd) [reply packages/web/src/components/editor.tsx:64]: Yes!

**Comment by tmhglnd** [comment]
Created a new pull request here https://github.com/Bubobubobubobubo/flok/pull/4

With fixes for the name badges and the unwanted dotted `...` line. I also added the `fontFamily` to the name badge so it changes with the selected font too. I took the liberty to lower the badge a bit as well, so it doesn't obstruct the code on that line too much.

**Comment by munshkr** [comment]
> Just a word of caution from personal experience: handling keyboard shortcuts can be quite difficult if you start to take into account that there will be multiple editing modes now. Some keyboard shortcuts will simply break because of the Vim Mode. Generally speaking, I'm not a big fan of these. A good command prompt can be tremendously better, especially since you only need to remember only one keybinding to summon it.

Also true, I guess we need to test them more. I performed a couple or days ago and had some issues with some shortcuts, for example, at one point, I tried to comment a line with CMD+/ but mistyped and pressed CMD+. and silenced everything 😨 same thing with Ctrl+Enter in Mac, which executes all the code instead of the current block.

> On another topic, I think that we are at a good point to start to get a bit creative with customization options. What about adding sliders for `font-weight` and other stuff?

Yes! That'd be great. We could make use of a Settings Dialog and throw everything in there 😄  I'd tackle this in a separate PR though

**Comment by Bubobubobubobubo** [comment]
Hello, bumping this PR. What can be done to get it merged in the near future?

**Comment by munshkr** [comment]
> Hello, bumping this PR. What can be done to get it merged in the near future?

Hey! Let me take a look at the to-do list I had, I don't quite remember if it's up to date, maybe some of these things can be done later and we can merge this now...

**Comment by munshkr** [comment]
Ok I think it's ready! Do you want to take a last look? Just in case you feel something is missing...

These were my last changes:
- Remove `font-size-adjust` rule. For some reason it made the font super small on my end. Maybe this is needed in some cases? @tmhglnd 
- Enable word wrapping by default
- Move editor settings into a single object, and save/load to/from local storage automatically.

The only thing that I didn't restored was the shortcuts for toggling Line numbers and Word wrapping, because I wasn't sure what was the best way to handle it now that it is a prop set in parent components and not handled in the editor itself. But I figured this isn't that bad now that settings are saved automatically, you just have to toggle them once if you prefer :)

**Comment by Bubobubobubobubo** [comment]
I don't have much time to program currently and I trust your judgment 🙏. IMO, the essential is to get something merged that we can gradually improve on. I'm sure more CSS hacks will find their way in in due time. I'll probably do more commits in the upcoming months when time will be a resource again.

**Comment by munshkr** [comment]
Alright, I'll merge it for now. As you mentioned, we can still make adjustments later. Thanks to both of you for your hard work and suggestions, and a shoutout to @diegodorado for motivating these customization improvements with the Vim mode! 😊

**Comment by tmhglnd** [comment]
> * Removed `font-size-adjust` rule. For some reason it made the font super small on my end. Maybe this is needed in some cases? @tmhglnd

I think I added this to reduce the font-size differences between the fonts. For me it looked good I remember, but if it gives issues on your end then we leave it out. I haven't had a look at it in a while. In any case I think all looks good so far! 

I don't mind not having the shortkeys for wrapping/line-numbers, the setting in the menu are fine.

**Comment by tmhglnd** [comment]
I also notice now that selection highlighting is difficult to see when the selection is just on 1 line. When switching to multiple lines it becomes better visible.

**Comment by diegodorado** [comment]
oh nice!

**Pull request merged**